File: Server/src/EffectsManager.ts
import { Probability } from "@spt-aki/models/eft/common/IGlobals"
import { AbstractModManager } from "./AbstractModManager"
import { Constants } from "./Constants"


export class EffectsManager extends AbstractModManager
{
    protected configName: string = "EffectsConfig"

    protected afterPostDB(): void
    {
        const effects = this.databaseTables.globals.config.Health.Effects

        if (this.config.probChangesEnabled)
        {
            this.setProbability(effects.LightBleeding.Probability, this.config.lightBleed.prob)
            this.setProbability(effects.HeavyBleeding.Probability, this.config.heavyBleed.prob)
            this.setProbability(effects.Fracture.FallingProbability, this.config.fracture.fallProb)
            this.setProbability(effects.Fracture.BulletHitProbability, this.config.fracture.bulletProb)
        }

        effects.LightBleeding.DamageHealth *= this.config.lightBleed.damageMult
        effects.LightBleeding.RemovePrice *= this.config.lightBleed.priceMult

        effects.HeavyBleeding.DamageHealth *= this.config.heavyBleed.damageMult
        effects.HeavyBleeding.RemovePrice *= this.config.heavyBleed.priceMult

        effects.Fracture.RemovePrice *= this.config.fracture.priceMult

        effects.Intoxication.DamageHealth *= this.config.toxin.damageMult
        effects.Intoxication.RemovePrice *= this.config.toxin.priceMult

        console.log(`${Constants.ModTitle}: Effects changes applied!`)
    }

    private setProbability(prob: Probability, config: any)
    {
        prob.Threshold = config.T
        prob.K = config.K
        prob.B = config.B
    }
}

File: Server/VAI-NRTP/Managers/EffectsManager.cs
using System.Text.Json.Nodes;
using SPTarkov.DI.Annotations;
using SPTarkov.Server.Core.Models.Eft.Common;
using SPTarkov.Server.Core.Models.Spt.Tables;

namespace VAI.NRTP.Managers;

[Injectable(InjectionType.Singleton)]
public sealed class EffectsManager : AbstractModManager
{
    protected override string ConfigName => "EffectsConfig";

    protected override void AfterPostDb()
    {
        var effects = DatabaseTables.Globals?.Configuration?.Health?.Effects;
        if (effects is null)
        {
            return;
        }

        var probChangesEnabled = GetConfigBool("probChangesEnabled");
        var lightBleed = GetConfigObject("lightBleed");
        var heavyBleed = GetConfigObject("heavyBleed");
        var fracture = GetConfigObject("fracture");
        var toxin = GetConfigObject("toxin");

        var lightBleeding = effects.LightBleeding;
        var heavyBleeding = effects.HeavyBleeding;
        var fractureEffect = effects.Fracture;
        var intoxication = effects.Intoxication;

        if (probChangesEnabled)
        {
            var lightProb = lightBleed?["prob"] as JsonObject;
            var heavyProb = heavyBleed?["prob"] as JsonObject;
            var fractureFallProb = fracture?["fallProb"] as JsonObject;
            var fractureBulletProb = fracture?["bulletProb"] as JsonObject;

            SetProbability(lightBleeding?.Probability, lightProb);
            SetProbability(heavyBleeding?.Probability, heavyProb);
            SetProbability(fractureEffect?.FallingProbability, fractureFallProb);
            SetProbability(fractureEffect?.BulletHitProbability, fractureBulletProb);
        }

        MultiplyMember(lightBleeding, GetNumber(lightBleed?["damageMult"]), (target, value) => target.DamageHealth *= value);
        MultiplyMember(lightBleeding, GetNumber(lightBleed?["priceMult"]), (target, value) => target.RemovePrice *= value);

        MultiplyMember(heavyBleeding, GetNumber(heavyBleed?["damageMult"]), (target, value) => target.DamageHealth *= value);
        MultiplyMember(heavyBleeding, GetNumber(heavyBleed?["priceMult"]), (target, value) => target.RemovePrice *= value);

        MultiplyMember(fractureEffect, GetNumber(fracture?["priceMult"]), (target, value) => target.RemovePrice *= value);

        MultiplyMember(intoxication, GetNumber(toxin?["damageMult"]), (target, value) => target.DamageHealth *= value);
        MultiplyMember(intoxication, GetNumber(toxin?["priceMult"]), (target, value) => target.RemovePrice *= value);

        Constants.GetLogger().Info($"{Constants.ModTitle}: Effects changes applied!");
    }

    private static void SetProbability(Probability? probability, JsonObject? config)
    {
        if (probability is null || config is null)
        {
            return;
        }

        var threshold = GetNumber(config["T"]);
        var k = GetNumber(config["K"]);
        var b = GetNumber(config["B"]);

        if (threshold != null)
        {
            probability.Threshold = threshold.Value;
        }

        if (k != null)
        {
            probability.K = k.Value;
        }

        if (b != null)
        {
            probability.B = b.Value;
        }
    }

    private static void MultiplyMember<T>(T? target, double? multiplier, Action<T, double> apply)
        where T : class
    {
        if (target is null || multiplier is null)
        {
            return;
        }

        apply(target, multiplier.Value);
    }

    private static double? GetNumber(JsonNode? node)
    {
        return GetNumberValue(node);
    }
}
